pkg/apis/batch/validation/validation.go

```diff
@@ -88,6 +88,16 @@ func ValidateJob(job *batch.Job, opts apivalidation.PodValidationOptions) field.
 	allErrs := apivalidation.ValidateObjectMeta(&job.ObjectMeta, true, apivalidation.ValidateReplicationControllerName, field.NewPath("metadata"))
 	allErrs = append(allErrs, ValidateGeneratedSelector(job)...)
 	allErrs = append(allErrs, ValidateJobSpec(&job.Spec, field.NewPath("spec"), opts)...)
+	if job.Spec.CompletionMode != nil && *job.Spec.CompletionMode == batch.IndexedCompletion && job.Spec.Completions != nil {
+		// For indexed job, the job controller appends a suffix (`-$INDEX`)
+		// to the pod hostname when indexed job create pods.
+		// The index could be maximum `.spec.completions-1`
+		// If we don't validate this here, the indexed job will fail to create pods later.
+		maximumPodHostname := fmt.Sprintf("%s-%d", job.ObjectMeta.Name, *job.Spec.Completions-1)
```

Comment:
completions can be 0:

https://github.com/kubernetes/kubernetes/blob/a5cf298a95b4a60789a6ba1caab931bdc980aa2d/pkg/apis/batch/validation/validation.go#L120-L122

https://github.com/kubernetes/kubernetes/blob/a5cf298a95b4a60789a6ba1caab931bdc980aa2d/staging/src/k8s.io/apimachinery/pkg/api/validation/generic.go#L81-L88

so completions - 1 would be -1?

Reply:
When the completions=0, there will be no pods created, thus no pod will have a hostname with index=-1 as suffix. 

But yes, there is one case with different result after judging `completions>0`. When the job name > 60 characters, it will cause error here as the length exceeds 63 after appending `--1`.

Maybe better to add restrict completions to be positive? (referred to https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs, there is a sentence “specify a **non-zero positive** value for .spec.completions.”) Besides, a job with `completions=0` seems useless.

Reply:
Since no pods will be created, you can just skip the check for completions = 0. But yes, I agree that completions=0 doesn't make sense. But that's not something we can change due to backwards compatibility.

Reply:
> When the completions=0, there will be no pods created, thus no pod will have a hostname with index=-1 as suffix.

then I'd either skip this check for completions=0 or set the minimum index we check length for to 0

Reply:
In the latest commit, this check is skipped when completions=0